fdbserver/masterserver.actor.cpp

```diff
@@ -250,6 +250,18 @@ struct MasterData : NonCopyable, ReferenceCounted<MasterData> {
 	// up-to-date in the presence of key range splits/merges.
 	VersionVector ssVersionVector;
 
+	// Captures versions that are committed, but whose previous versions aren't committed yet.
+	// Each entry in this map is a (prevVersion, commitVersion) pair.
+	std::unordered_map<Version, Version> queuedCommittedVersions; // prevVersion -> commitVersion
```

Comment:
Yes, I went back and forth on this - I was concerned about the cost of maintaining these entries in sorted order (since I wasn't sure how many entries this map would have in a realistic scenario). Dan's (YCSB) test run showed a maximum of 50 entries (also, not that many instances with 50 entries) so it is probably fine to change it to "std::map". 
>since then we can use lower_bound and upper_bound

Each of these would require an iteration over the entries of the map, and that's what I tried to avoid here. Each of the entries in the map can be treated as an interval ([prevVersion, committedVersion]) and this code is essentially doing an interval merge (combining the current interval with the intervals that are already present, if any, in the map) in a single iteration - the downside of that is the code complexity.

Dan suggested using an interval management boost library API, if such an API is available, for this. I see that FoundationDB does use boost library, so that's something we can pursue.

Reply:
Also, as I mentioned to Dan this morning, if we don't try to merge intervals (or the entries in the map) then the code will be much simpler. For example, if we have entries [a, b] and [c, d] in the map and when we receive a committed version [b, c] (prevVersion = b, committedVersion = c) we can simply add it to the map, resulting in entries [a, b], [b, c], and [c, d] in the map - the code in this PR will combine these entries into a single entry [a, d]. If we don't have a whole lot of out-of-order commits then that's something that we can pursue.

Reply:
[here](https://www.boost.org/doc/libs/1_46_1/libs/icl/doc/html/index.html) is a possible library that could be used.